Parsing logic: fail gracefully on unexpected input

franzpoeschel opened the pull request:
Reading failures should ideally affect only the innermost scope possible. When one record component fails parsing due to whatever error, the neighboring record components should remain readable. When one iteration fails parsing, skip it and go on to the next.
Current behavior: Parsing fails and the Series cannot be opened

Close #983 

TODO
- [x] Introduce `ReadError` error type that backends can throw to tell the frontend about recoverable failures
- [x] Throw `ReadError` inside ADIOS1 and ADIOS2 attribute reading routines
- [x] Frontend: Recover from `ReadError`s ~~(partially implemented so far)~~
- [x] ~~Other backends,~~ also consider dataset reading, ... UPDATE: Catching wrong attribute reads now possible in all backends, dataset reading and further things can come in a later update
- [x] Testing: ~~file-based, group-based, varialbe-based, deferred/eager~~, opening single iterations of a file-based series, ~~streaming API, skipping iteration~~
- [x] Merge #949 first to avoid merge conflicts
- [x] Correctly recover from deferred actions in ADIOS2
- [x] ADIOS2 / variable-based encoding: Backend reports wrong value for snapshot in iteration 10
- [x] ~~bp4_steps test seems to not write snapshot attribute~~ only written in new ADIOS2 schema as it should be
- [x] Avoid const_cast by using the parsing state

This whole PR is a bit of a stretch-goal as there are many different types and sources of errors to be recovered from.
@ax3l Can you name some specific error pattern (or even give me a test file) that the openPMD-api should handle more robustly?

I think that a backend failure should mean that the entire call to `AbstractIOHandler::flush()` has failed and the queue is cleared. For finer granularity error handling, the frontend must issue single `flush()` calls. So, I've changed error recovery procedures to clear the entire IO queue upon failure.

franzpoeschel commented on src/ParticleSpecies.cpp:
```diff
@@ -111,12 +111,11 @@ void ParticleSpecies::read()
-            std::cerr << "WARNING: Skipping invalid openPMD record '"
-                      << record_name << "'" << std::endl;
-            while (!IOHandler()->m_work.empty())
-                IOHandler()->m_work.pop();
```
No need to clear the queue here anymore, it's already cleared by `AbstractIOHandler::flush()` upon failure

franzpoeschel commented on src/Iteration.cpp:
```diff
@@ -327,26 +336,6 @@ void Iteration::deferParseAccess(DeferredParseAccess dr)
         std::make_optional<DeferredParseAccess>(std::move(dr));
 }
 
-void Iteration::read()
```
Merged this with `runDeferredParseAccess`, no reason to keep them separate.

franzpoeschel commented on src/Iteration.cpp:
```diff
@@ -576,7 +595,15 @@ AdvanceStatus Iteration::beginStep()
+        {
+            *newType = oldType;
+            throw;
+        }
```
This style of try-catch-block is necessary if we want to keep the Series usable after throwing an exception. If we want to avoid this, we must find a way to parse without changing that state.

franzpoeschel commented:
(not actually ready for review, just wanting to run the CI)

franzpoeschel commented on src/IO/HDF5/HDF5IOHandler.cpp:
```diff
@@ -751,9 +751,15 @@ void HDF5IOHandlerImpl::openPath(
-    VERIFY(
-        node_id >= 0,
-        "[HDF5] Internal error: Failed to open HDF5 group during path opening");
+    if (node_id <= 0)
```
```suggestion
    if (node_id < 0)
```

franzpoeschel commented on src/IO/HDF5/HDF5IOHandler.cpp:
```diff
@@ -764,30 +770,50 @@ void HDF5IOHandlerImpl::openPath(
-            path_id >= 0,
-            "[HDF5] Internal error: Failed to open HDF5 group during path "
-            "opening");
+        if (path_id <= 0)
```
```suggestion
        if (path_id < 0)
```

franzpoeschel commented on src/IO/HDF5/HDF5IOHandler.cpp:
```diff
@@ -1674,10 +1712,15 @@ void HDF5IOHandlerImpl::readAttribute(
-        status == ndims,
-        "[HDF5] Internal error: Failed to get dimensions during attribute "
-        "read");
+    if (obj_id < 0)
```
```suggestion
    if (status != ndims)
```

ax3l commented:
> Can you name some specific error pattern (or even give me a test file) that the openPMD-api should handle more robustly?
 
The most important pattern I have it reading a series (any encoding) from a sim that is currently running. The latest time step is usually incomplete.
I think we could create test files by aborting sims mid-dump somehow...?

franzpoeschel committed:
Add ReadError error type

franzpoeschel committed:
Backend additions: ADIOS2

franzpoeschel committed:
Backend additions: ADIOS1

franzpoeschel committed:
Backend additions: HDF5

franzpoeschel committed:
Backend additions: JSON

franzpoeschel committed:
Fully clear task queue upon failure

franzpoeschel commented:
I've now reproduced the only remaining [CI failure](https://github.com/openPMD/openPMD-api/runs/6708072907?check_suite_focus=true) locally. It looks like for some reason Clang6 does not catch an exception despite there being a fitting catch block. Not sure what to do with this yet.

franzpoeschel commented:
If I add `catch (...)` at the offending places, everything works fine again. For some reason, clang-6 loses information about exception types across static libraries.

This same workaround is not necessary for the other backends.

franzpoeschel commented:
It looks like Clang with libc++ has trouble catching exceptions: https://stackoverflow.com/questions/51673343/clang-with-libc-exceptions
This only affects ADIOS1 in our case, the rest is working. I think we should either disable that test in this CI workflow, or disable ADIOS1 for that workflow.

Relevant:
![Bildschirmfoto vom 2022-06-17 14-09-01](https://user-images.githubusercontent.com/14241876/174300103-1cd429fc-dd57-4af2-bbb6-ac5fa986fbfd.png)

franzpoeschel commented on CMakeLists.txt:
```diff
@@ -484,21 +484,13 @@ set(IO_SOURCE
-        src/ChunkInfo.cpp
         src/IO/ADIOS/CommonADIOS1IOHandler.cpp
-        src/IO/ADIOS/ParallelADIOS1IOHandler.cpp
-        src/IO/IOTask.cpp)
```
I don't fully understand why these need to be part of the ADIOS1 backend libs. Shouldn't it be enough to have the compiled symbols in libopenPMD.so and link later?
I removed them for now to reduce the number of files that must not include `Error.hpp`. If there are any .cpp files that should be added back here, we can set `#define OPENPMD_ADIOS1_IMPLEMENTATION 1` at its head. Error.hpp will throw a compile-time error if it finds that macro.

franzpoeschel replied:
@ax3l can you have a look at this, please?

ax3l replied:
I think we need the symbols in these libs because the openPMD-api adios1 wrapper libs are not depending on the main openPMD-api libs (but the other way around).

Afaik, that only manifests as an issue in macOS. Maybe we can build a mini-app in macOS CI on our installed lib to cover this.

franzpoeschel replied:
We should definitely introduce such a CI check, as the error symbol workaround in this PR introduces a reverse dependency, where the ADIOS1 backend needs symbols from the main library. 
In the end, we might need to restrict the workaround to non-Apple systems.
I'll push a commit that reverses the changes in here.

franzpoeschel commented:
I think I was now able to avoid the incompatibility issues by removing `Error.hpp` from the ADIOS1 backend targets entirely, replacing them with functions that throw the errors, e.g.:
```cpp
    void throwReadError(
        AffectedObject affectedObject,
        Reason reason,
        std::optional<std::string> backend,
        std::string description)
    {
        throw ReadError(
            affectedObject, reason, std::move(backend), std::move(description));
    }
```

This way, libopenPMD.ADIOS1.(Serial|Parallel) don't define the error symbols at all.

franzpoeschel committed:
Move error throwing outside of ADIOS1 library

Some weird compiler configurations don't understand error type symbols
across libraries, so keep the symbols entirely to the main library.

franzpoeschel committed:
Error handling at Series level

This is the commit with the most complicated logic changes.
It implements skipping broken iterations at the Series level.
This includes the Streaming API (Series::readIterations()), as well as
our three iteration encodings.

On the positive side: The most complex logic changes have already been
prepared in the topic-adios2-append PR

franzpoeschel committed:
Error handling throughout the openPMD object model

Nothing too complicated, just tedious

franzpoeschel committed:
Testing, and adapt tests to new Error types

franzpoeschel committed:
Replace no_such_file_error by ReadError

`using no_such_file_error = error::ReadError` for backwards
compatibility in user code.

franzpoeschel committed:
Forward errors instead of creating new ones

@todo: Check if this needs to be done elsewhere

franzpoeschel committed:
Test opening single broken filebased iterations

ax3l commented:
Thanks  a lot!

I did a first test reading partially written simulation data, i.e., the `/openPMD` attribute is not yet written for each iteration. I tested it in `fileBased` encoding. Currently, incomplete iterations (snapshots) are not yet leniently skipped:
```
openpmd-ls ~/src/warpx/build/bin/diags/diag_btd/openpmd_%T.bp
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
[~ADIOS2IOHandler] An error occurred: [ADIOS2] Requested attribute (/openPMD) not found in backend.
An error occurred while opening the specified openPMD series!
[ADIOS2] Requested attribute (/openPMD) not found in backend.
```
[partial_btd.tar.gz](https://github.com/openPMD/openPMD-api/files/10259146/partial_btd.tar.gz)

franzpoeschel commented:
This is the output that I get upon `openpmd-ls openpmd_%T.bp` with your dataset:

```
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '0' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '1' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '2' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '3' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '4' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '5' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '6' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '7' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '8' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '9' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '10' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '11' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '12' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '13' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '14' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '15' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '16' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '17' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '18' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '19' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '20' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '21' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '22' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '23' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '24' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '25' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '26' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '27' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '28' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '29' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '30' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '31' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '32' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '33' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '34' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '35' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '36' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '37' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '38' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '39' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '40' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '41' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '42' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '43' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '44' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '45' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '46' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '47' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '48' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '49' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '50' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '51' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '52' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '53' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '54' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '55' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '56' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '57' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '58' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '59' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '60' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '61' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '62' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '63' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '64' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '65' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '66' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '67' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '68' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '69' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '70' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '71' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '72' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '73' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '74' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '75' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '76' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '77' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '78' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '79' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '80' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '81' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '82' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '83' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '84' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '85' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '86' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '87' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '88' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '89' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '90' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '91' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '92' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '93' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '94' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '95' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '96' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '97' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '98' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[ADIOS2] Warning: Attribute with name /openPMD has no type in backend.
Cannot read iteration '99' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:	Attribute
Error type:	NotFound
Further description:	/openPMD
[fwk394:1588514:0:1588514] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x6)
==== backtrace (tid:1588514) ====
 0 0x000000000003dbf0 __GI___sigaction()  ???:0
 1 0x000000000046559c std::__1::__sort<openPMD::ADIOS2IOHandlerImpl::~ADIOS2IOHandlerImpl()::$_2&, std::__1::unique_ptr<openPMD::detail::BufferedActions, std::__1::default_delete<openPMD::detail::BufferedActions> >*>()  ADIOS2IOHandler.cpp:0
 2 0x0000000000465475 std::__1::__sort<openPMD::ADIOS2IOHandlerImpl::~ADIOS2IOHandlerImpl()::$_2&, std::__1::unique_ptr<openPMD::detail::BufferedActions, std::__1::default_delete<openPMD::detail::BufferedActions> >*>()  ADIOS2IOHandler.cpp:0
 3 0x0000000000465475 std::__1::__sort<openPMD::ADIOS2IOHandlerImpl::~ADIOS2IOHandlerImpl()::$_2&, std::__1::unique_ptr<openPMD::detail::BufferedActions, std::__1::default_delete<openPMD::detail::BufferedActions> >*>()  ADIOS2IOHandler.cpp:0
 4 0x000000000045202e openPMD::ADIOS2IOHandlerImpl::~ADIOS2IOHandlerImpl()  ???:0
 5 0x000000000046fc76 openPMD::ADIOS2IOHandler::~ADIOS2IOHandler()  ???:0
 6 0x000000000046fde9 openPMD::ADIOS2IOHandler::~ADIOS2IOHandler()  ???:0
 7 0x00000000002f2cce openPMD::internal::SeriesData::close()  ???:0
 8 0x00000000002f282e openPMD::internal::SeriesData::~SeriesData()  ???:0
 9 0x00000000002f2d69 openPMD::internal::SeriesData::~SeriesData()  ???:0
10 0x0000000000287bd5 openPMD::Attributable::~Attributable()  ???:0
11 0x00000000002f3e2b openPMD::Series::Series()  ???:0
12 0x00000000000c7a0e pybind11::error_already_set::m_fetched_error_deleter()  ???:0
13 0x000000000002f95a pybind11::error_already_set::m_fetched_error_deleter()  ???:0
14 0x0000000000129fc3 cfunction_call()  ???:0
15 0x00000000000bcc84 _PyObject_MakeTpCall.localalias()  ???:0
16 0x000000000006f41c _PyEval_EvalFrameDefault()  ???:0
17 0x00000000001eb2cf _PyEval_Vector()  ???:0
18 0x000000000006d68f _PyEval_EvalFrameDefault()  ???:0
19 0x00000000001eb2cf _PyEval_Vector()  ???:0
20 0x00000000001eb938 PyEval_EvalCode()  ???:0
21 0x000000000026f8ad run_mod()  ???:0
22 0x000000000027c382 _PyRun_SimpleFileObject.localalias()  ???:0
23 0x000000000027c95b _PyRun_AnyFileObject()  ???:0
24 0x0000000000280a1f Py_RunMain.localalias()  ???:0
25 0x0000000000281245 Py_BytesMain()  ???:0
26 0x000000000002924e __libc_start_call_main()  ???:0
27 0x0000000000029309 __libc_start_main_alias_2()  ???:0
28 0x0000000000401075 _start()  ???:0
=================================
```

Weirdly enough, the segfault only appears when redirecting stderr into a file....

update: we use `std::stable_sort` not in the destructor, even though there are no duplicate files.

franzpoeschel committed:
Use std::stable_sort in ADIOS2 destructor

ax3l commented on CMakeLists.txt:
```diff
@@ -909,6 +907,7 @@ if(openPMD_USE_INVASIVE_TESTS)
         message(WARNING "Invasive tests that redefine class signatures are "
                         "known to fail on Windows!")
     endif()
+    target_compile_definitions(openPMD PRIVATE openPMD_USE_INVASIVE_TESTS=1)
```
This macro should never be shipped. It introduces UB and is only used in some testing.
```suggestion
```

franzpoeschel replied:
The macro is still activated only in those special tests and not shipped (see the `if(openPMD_USE_INVASIVE_TESTS)` above). I need this macro inside `Iteration.cpp` in here: https://github.com/openPMD/openPMD-api/blob/8c0d2a21dd3945523dad5af7cd07b380592cd6d2/src/Iteration.cpp#L542
This is the most reliable way to trigger errors to be thrown.

Are there workflows where users compile openPMD with `openPMD_USE_INVASIVE_TESTS` and still install the resulting compiled library? If yes, are there better ways to do this then? 
Just removing this line breaks the parsing test.

ax3l replied:
ah, I see.
ok, extents its meaning but is ok.

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -112,7 +112,17 @@ ADIOS2IOHandlerImpl::~ADIOS2IOHandlerImpl()
+     * Technically, std::sort() is sufficient here, since file names are unique.
+     * Use std::stable_sort() for two reasons:
+     * 1) On some systems (clang 13.0.1, libc++ 13.0.1), std::sort() leads to
+     *    weird inconsisten segfaults here.
```
```suggestion
     *    weird inconsistent segfaults here.
```

ax3l commented on src/ParticleSpecies.cpp:
```diff
@@ -76,7 +86,20 @@
+                          << err.what() << std::endl;
+
+                map.forget(record_name);
+                //(*this)[record_name].erase(RecordComponent::SCALAR);
+                // this->erase(record_name);
```
ping @franzpoeschel let's clean?

franzpoeschel committed:
Add removed .cpp files back to ADIOS1 backend implementation

ax3l commented on src/RecordComponent.cpp:
```diff
@@ -350,7 +362,11 @@ void RecordComponent::readBase()
+            error::AffectedObject::Attribute,
+            error::Reason::UnexpectedContent,
+            {},
+            "Unexpected Attribute datatype for 'unitSI'");
```
For such errors, maybe we should write also what we expect and what we got in the error message?
```suggestion
            "Unexpected Attribute datatype for 'unitSI' (expected: double; found: ...)");
```

ax3l commented on src/Series.cpp:
```diff
@@ -1017,38 +1021,144 @@ void Series::readFileBased()
+                    error::Reason::Other,
+                    {},
+                    "Not a single iteration can be successfully parsed (see "
+                    "above "
+                    "warnings).");
```
```suggestion
                    "above warnings).");
```

ax3l reviewed (commented):
This looks great! 
Added inline comments :)

franzpoeschel committed:
Code review

franzpoeschel committed:
Add export attribute to functions in ThrowError.hpp

ax3l reviewed (approved):
:rocket: :sparkles:
